File: java/com/google/copybara/onboard/ConfigHeuristicsInputProvider.java
/*
 * Copyright (C) 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.onboard;

import static com.google.common.collect.ImmutableList.toImmutableList;
import static com.google.common.collect.ImmutableSet.toImmutableSet;

import com.google.common.base.Ascii;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import com.google.copybara.GeneralOptions;
import com.google.copybara.configgen.ConfigGenHeuristics;
import com.google.copybara.configgen.ConfigGenHeuristics.DestinationExcludePaths;
import com.google.copybara.configgen.ConfigGenHeuristics.GeneratorTransformations;
import com.google.copybara.configgen.ConfigGenHeuristics.Result;
import com.google.copybara.exception.RepoException;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.git.FuzzyClosestVersionSelector;
import com.google.copybara.git.GitOptions;
import com.google.copybara.git.GitRepository;
import com.google.copybara.git.GitRevision;
import com.google.copybara.onboard.core.CannotProvideException;
import com.google.copybara.onboard.core.Input;
import com.google.copybara.onboard.core.InputProvider;
import com.google.copybara.onboard.core.InputProviderResolver;
import com.google.copybara.remotefile.HttpStreamFactory;
import com.google.copybara.remotefile.RemoteFileOptions;
import com.google.copybara.remotefile.extractutil.ExtractType;
import com.google.copybara.remotefile.extractutil.ExtractUtil;
import com.google.copybara.util.Glob;
import com.google.copybara.util.console.Console;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Stream;

/**
 * An input provider that uses the origin and destination content information to infer several
 * fields like the origin_files glob.
 */
public class ConfigHeuristicsInputProvider implements InputProvider {

  private static final Glob INCLUDE_EXCLUDE_NOOP =
      Glob.createGlob(ImmutableList.of("**"), ImmutableList.of("**"));

  @SuppressWarnings({"OptionalUsedAsFieldOrParameterType", "OptionalAssignedToNull"})
  private Optional<Result> cached = null;

  private CannotProvideException cachedException = null;

  private final GitOptions gitOptions;
  private final GeneralOptions generalOptions;
  private final GeneratorOptions generatorOptions;
  private final RemoteFileOptions remoteFileOptions;
  private final ImmutableSet<Path> destinationOnlyPaths;
  private final int percentSimilar;
  private final Console console;
  private final DestinationPathProvider destinationPathProvider;

  public ConfigHeuristicsInputProvider(
      GitOptions gitOptions,
      GeneralOptions generalOptions,
      GeneratorOptions generatorOptions,
      RemoteFileOptions remoteFileOptions,
      ImmutableSet<Path> destinationOnlyPaths,
      int percentSimilar,
      Console console,
      DestinationPathProvider destinationPathProvider) {
    this.gitOptions = gitOptions;
    this.generalOptions = generalOptions;
    this.generatorOptions = generatorOptions;
    this.remoteFileOptions = remoteFileOptions;
    this.destinationOnlyPaths = destinationOnlyPaths;
    this.percentSimilar = percentSimilar;
    this.console = console;
    this.destinationPathProvider = destinationPathProvider;
  }

  @Override
  public <T> Optional<T> resolve(Input<T> input, InputProviderResolver db)
      throws InterruptedException, CannotProvideException {
    Path destination = destinationPathProvider.resolve(db);
    Optional<Result> result = computeHeuristic(db, destination);
    if (result.isEmpty()) {
      return Optional.empty();
    }
    if (input == Inputs.ORIGIN_GLOB) {
      Glob resultGlob = result.get().getOriginGlob();
      return resultGlob.equals(INCLUDE_EXCLUDE_NOOP)
          ? Optional.empty()
          : Optional.of(Inputs.ORIGIN_GLOB.asValue(resultGlob));
    }
    if (input == Inputs.TRANSFORMATIONS) {
      GeneratorTransformations transformations = result.get().getTransformations();
      return Optional.of(Inputs.TRANSFORMATIONS.asValue(transformations));
    }
    if (input == Inputs.DESTINATION_EXCLUDE_PATHS) {
      DestinationExcludePaths destinationExcludePaths = result.get().getDestinationExcludePaths();
      boolean optimizeGlobs = generatorOptions.optimizeGlobs;
      if (optimizeGlobs) {
        // We filter out paths containing "AUTOPATCHES" unconditionally here. This optimizes the
        // generated globs by excluding autopatch paths, which are intended to be managed by
        // Copybara rather than explicitly listed as excludes.
        ImmutableSet<Path> filtered =
            destinationExcludePaths.getPaths().stream()
                .filter(p -> !p.toString().contains("AUTOPATCHES"))
                .collect(toImmutableSet());
        destinationExcludePaths = new DestinationExcludePaths(filtered);
      }
      return Optional.of(Inputs.DESTINATION_EXCLUDE_PATHS.asValue(destinationExcludePaths));
    }
    return Optional.empty();
  }

  @SuppressWarnings("OptionalAssignedToNull")
  protected Optional<Result> computeHeuristic(InputProviderResolver db, Path destination)
      throws InterruptedException, CannotProvideException {
    if (cachedException != null) {
      throw cachedException;
    }
    if (!Files.isDirectory(destination)) {
      return Optional.empty();
    }
    if (cached != null) {
      return cached;
    }

    String originForLog = "unknown";
    try {
      Path origin = generalOptions.getDirFactory().newTempDir("checkout");
      ImmutableList<String> upstreamTags = ImmutableList.of();

      OriginUrls urls = resolveOriginUrls(db);
      URL archiveUrl = urls.archiveUrl;
      URL originUrl = urls.originUrl;

      if (archiveUrl != null) {
        originForLog = archiveUrl.toString();
        String unpackMethod = db.resolve(Inputs.UNPACK_METHOD);
        console.infoFmt("Downloading archive from %s", archiveUrl);
        HttpStreamFactory transport = remoteFileOptions.getTransport();
        try (InputStream inputStream = transport.open(archiveUrl, null)) {
          ExtractType extractType;
          try {
            extractType = ExtractType.valueOf(Ascii.toUpperCase(unpackMethod));
          } catch (IllegalArgumentException e) {
            throw new ValidationException(
                String.format("Invalid unpack method '%s'", unpackMethod), e);
          }
          ExtractUtil.extractArchive(inputStream, origin, extractType, null);
        }

        // Verify that the archive was extracted successfully.
        long fileCount = 0;
        try (Stream<Path> stream = Files.walk(origin)) {
          fileCount = stream.filter(Files::isRegularFile).count();
        } catch (IOException e) {
          console.warnFmt("Failed to iterate files in extracted archive: %s", e.getMessage());
        }
        console.infoFmt("Extracted %d files from remote archive", fileCount);
        if (fileCount == 0) {
          console.warnFmt(
              "Remote archive yielded 0 files! This might indicate download/unpack failure.");
        }
      } else if (originUrl != null) {
        originForLog = originUrl.toString();
        String currentVersion = db.resolve(Inputs.CURRENT_VERSION);

        GitRepository repo =
            gitOptions.cachedBareRepoForUrl(originUrl.toString()).withWorkTree(origin);

        FuzzyClosestVersionSelector selector = new FuzzyClosestVersionSelector();
        currentVersion =
            selector.selectVersion(currentVersion, repo, originUrl.toString(), console);

        console.progressFmt("Fetching '%s' from %s", currentVersion, originUrl.toString());
        GitRevision gitRevision;
        try {
          gitRevision =
              repo.fetchSingleRefWithTags(
                  originUrl.toString(),
                  currentVersion,
                  /* fetchTags= */ true,
                  /* partialFetch= */ false,
                  Optional.empty());
        } catch (RepoException e) {
          gitRevision =
              repo.fetchSingleRef(
                  originUrl.toString(),
                  currentVersion,
                  /* partialFetch= */ false,
                  Optional.empty());
        }
        Path git = Files.createDirectories(origin);
        upstreamTags =
            repo.showRef().keySet().stream()
                .filter(ref -> ref.startsWith("refs/tags/"))
                .collect(toImmutableList());

        console.progressFmt("Checking out git files");
        repo.withWorkTree(git).forceCheckout(gitRevision.getHash());
      } else {
        throw new CannotProvideException(
            "Neither GIT_ORIGIN_URL nor REMOTE_ARCHIVE_URL was provided.");
      }

      ConfigGenHeuristics heuristics =
          getConfigGenHeuristics(
              destination,
              origin,
              destinationOnlyPaths,
              percentSimilar,
              generatorOptions,
              generalOptions,
              upstreamTags);

      console.progressFmt("Computing globs");
      cached = Optional.of(heuristics.run());
      return cached;

    } catch (ValidationException | IOException | RepoException e) {
      cachedException =
          new CannotProvideException(
              String.format(
                  "Cannot compute heuristics for repository %s: %s", originForLog, e.getMessage()),
              e);
      throw cachedException;
    }
  }

  /** Represents the resolved origin URLs. */
  protected static class OriginUrls {
    public final URL archiveUrl;
    public final URL originUrl;

    public OriginUrls(URL archiveUrl, URL originUrl) {
      this.archiveUrl = archiveUrl;
      this.originUrl = originUrl;
    }
  }

  /** Resolves the origin URLs from the resolver. */
  protected OriginUrls resolveOriginUrls(InputProviderResolver db)
      throws InterruptedException, CannotProvideException {
    // 1. Peek silently to see if a Git URL is already known (e.g. from flags or file).
    URL originUrl = db.resolveOptional(Inputs.GIT_ORIGIN_URL).orElse(null);
    URL archiveUrl = null;

    // 2. If no Git URL found, peek silently for an Archive URL.
    if (originUrl == null) {
      archiveUrl = db.resolveOptional(Inputs.REMOTE_ARCHIVE_URL).orElse(null);
    }

    // 3. If still nothing found silently, force resolution (may prompt user) based on template.
    if (originUrl == null && archiveUrl == null) {
      if (Objects.equals(generatorOptions.getTemplate(), "remote_archive_to_third_party")) {
        archiveUrl = db.resolve(Inputs.REMOTE_ARCHIVE_URL);
      } else {
        originUrl = db.resolve(Inputs.GIT_ORIGIN_URL);
      }
    }
    return new OriginUrls(archiveUrl, originUrl);
  }

  /**
   * Returns a {@link ConfigGenHeuristics} object.
   *
   * @param destination the local path to the destination
   * @param origin the local path to the origin
   * @param destinationOnlyPaths paths that should be considered destination-only, and excluded from
   *     heuristics
   * @param percentSimilar the threshold for considering an origin file similar enough to a
   *     destination file
   * @param generatorOptions the generator options from {@link com.google.copybara.Options}
   * @param generalOptions the general options from {@link com.google.copybara.Options}
   * @param versions a list of version refs from the upstream
   * @return the object
   */
  protected ConfigGenHeuristics getConfigGenHeuristics(
      Path destination,
      Path origin,
      ImmutableSet<Path> destinationOnlyPaths,
      int percentSimilar,
      GeneratorOptions generatorOptions,
      GeneralOptions generalOptions,
      ImmutableList<String> versions) {
    return new ConfigGenHeuristics(
        origin,
        destination,
        destinationOnlyPaths,
        percentSimilar,
        generatorOptions.computeGlobIgnoreCarriageReturn,
        generatorOptions.computeGlobIgnoreWhitespace,
        generalOptions,
        versions);
  }

  @Override
  public ImmutableMap<Input<?>, Integer> provides() throws CannotProvideException {
    return defaultPriority(
        ImmutableSet.of(
            Inputs.ORIGIN_GLOB, Inputs.TRANSFORMATIONS, Inputs.DESTINATION_EXCLUDE_PATHS));
  }

  /**
   * Resolves a destination path for glob generation heuristics. This allows the destination path to
   * be different than the generator output folder, if needed.
   */
  @FunctionalInterface
  public interface DestinationPathProvider {
    Path resolve(InputProviderResolver db) throws InterruptedException, CannotProvideException;
  }
}

File: src/Copybara.Core/Onboard/ConfigHeuristicsInputProvider.cs
/*
 * Copyright (C) 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;

using Copybara.ConfigGen;
using Copybara.Exceptions;
using Copybara.Git;
using Copybara.Onboard.Core;
using Copybara.Util;

using Console = Copybara.Util.Console.Console;

namespace Copybara.Onboard;

/// <summary>
/// An input provider that uses the origin and destination content information to infer several
/// fields like the origin_files glob. Port of
/// <c>com.google.copybara.onboard.ConfigHeuristicsInputProvider</c>.
///
/// <para><b>Port note:</b> depends on <c>Copybara.ConfigGen.ConfigGenHeuristics</c> (ported by a
/// peer). Java's <c>java.net.URL</c> maps to <see cref="System.Uri"/> and <c>java.nio.file.Path</c>
/// to <c>string</c>.</para>
/// </summary>
public class ConfigHeuristicsInputProvider : IInputProvider
{
    private static readonly Glob IncludeExcludeNoop =
        Glob.CreateGlob(ImmutableArray.Create("**"), ImmutableArray.Create("**"));

    // Optional sentinel: null means "not computed yet"; a null-valued result means "no result".
    private bool _computed;
    private ConfigGenHeuristics.Result? _cached;

    private readonly GitOptions _gitOptions;
    private readonly GeneralOptions _generalOptions;
    private readonly GeneratorOptions _generatorOptions;
    private readonly ImmutableHashSet<string> _destinationOnlyPaths;
    private readonly int _percentSimilar;
    private readonly Console _console;
    private readonly DestinationPathProvider _destinationPathProvider;

    public ConfigHeuristicsInputProvider(
        GitOptions gitOptions,
        GeneralOptions generalOptions,
        GeneratorOptions generatorOptions,
        ImmutableHashSet<string> destinationOnlyPaths,
        int percentSimilar,
        Console console,
        DestinationPathProvider destinationPathProvider)
    {
        _gitOptions = gitOptions;
        _generalOptions = generalOptions;
        _generatorOptions = generatorOptions;
        _destinationOnlyPaths = destinationOnlyPaths;
        _percentSimilar = percentSimilar;
        _console = console;
        _destinationPathProvider = destinationPathProvider;
    }

    public T? Resolve<T>(Input<T> input, IInputProviderResolver db)
        where T : class
    {
        Uri originUrl = db.Resolve(Inputs.GitOriginUrl);
        string currentVersion = db.Resolve(Inputs.CurrentVersion);
        string destination = _destinationPathProvider(db);
        ConfigGenHeuristics.Result? result = ComputeHeuristic(originUrl, currentVersion, destination);
        if (result == null)
        {
            return null;
        }

        if (ReferenceEquals(input, Inputs.OriginGlob))
        {
            Glob resultGlob = result.GetOriginGlob();
            return resultGlob.Equals(IncludeExcludeNoop)
                ? null
                : (T?)(object)Inputs.OriginGlob.AsValue(resultGlob);
        }

        if (ReferenceEquals(input, Inputs.Transformations))
        {
            ConfigGenHeuristics.GeneratorTransformations transformations = result.GetTransformations();
            return (T?)(object)Inputs.Transformations.AsValue(transformations);
        }

        if (ReferenceEquals(input, Inputs.DestinationExcludePaths))
        {
            ConfigGenHeuristics.DestinationExcludePaths destinationExcludePaths =
                result.GetDestinationExcludePaths();
            if (_generatorOptions.OptimizeGlobs)
            {
                // Filter out paths containing "AUTOPATCHES" unconditionally here. This optimizes the
                // generated globs by excluding autopatch paths, which are intended to be managed by
                // Copybara rather than explicitly listed as excludes.
                ImmutableHashSet<string> filtered =
                    destinationExcludePaths.GetPaths()
                        .Where(p => !p.Contains("AUTOPATCHES"))
                        .ToImmutableHashSet();
                destinationExcludePaths = new ConfigGenHeuristics.DestinationExcludePaths(filtered);
            }

            return (T?)(object)Inputs.DestinationExcludePaths.AsValue(destinationExcludePaths);
        }

        return null;
    }

    protected ConfigGenHeuristics.Result? ComputeHeuristic(
        Uri originUrl, string currentVersion, string destination)
    {
        if (!Directory.Exists(destination))
        {
            return null;
        }

        if (_computed)
        {
            return _cached;
        }

        try
        {
            // TODO(malcon): Refactor this class to not depend on git. IOW, be able to generate
            // configs for existing sources for non-git repositories.
            string origin = _generalOptions.GetDirFactory().NewTempDir("checkout");
            GitRepository repo =
                _gitOptions.CachedBareRepoForUrl(originUrl.ToString()).WithWorkTree(origin);

            var selector = new FuzzyClosestVersionSelector();
            currentVersion =
                selector.SelectVersion(currentVersion, repo, originUrl.ToString(), _console);

            _console.ProgressFmt("Fetching '%s' from %s", currentVersion, originUrl.ToString());
            GitRevision gitRevision;
            try
            {
                gitRevision =
                    repo.FetchSingleRefWithTags(
                        originUrl.ToString(),
                        currentVersion,
                        fetchTags: true,
                        partialFetch: false,
                        null);
            }
            catch (RepoException)
            {
                gitRevision =
                    repo.FetchSingleRef(
                        originUrl.ToString(), currentVersion, partialFetch: false, null);
            }

            Directory.CreateDirectory(origin);
            string git = origin;
            ImmutableArray<string> upstreamTags =
                repo.ShowRef().Keys
                    .Where(r => r.StartsWith("refs/tags/", StringComparison.Ordinal))
                    .ToImmutableArray();

            _console.ProgressFmt("Checking out git files");
            repo.WithWorkTree(git).ForceCheckout(gitRevision.GetHash());

            ConfigGenHeuristics heuristics =
                GetConfigGenHeuristics(
                    destination,
                    origin,
                    _destinationOnlyPaths,
                    _percentSimilar,
                    _generatorOptions,
                    _generalOptions,
                    upstreamTags);

            _console.ProgressFmt("Computing globs");
            _cached = heuristics.Run();
            _computed = true;
            return _cached;
        }
        catch (Exception e) when (e is ValidationException or IOException or RepoException)
        {
            // Cannot compute heuristics for this repository.
            _cached = null;
            _computed = true;
            return _cached;
        }
    }

    /// <summary>Returns a <see cref="ConfigGenHeuristics"/> object.</summary>
    protected ConfigGenHeuristics GetConfigGenHeuristics(
        string destination,
        string origin,
        ImmutableHashSet<string> destinationOnlyPaths,
        int percentSimilar,
        GeneratorOptions generatorOptions,
        GeneralOptions generalOptions,
        ImmutableArray<string> versions) =>
        new(
            origin,
            destination,
            destinationOnlyPaths,
            percentSimilar,
            generatorOptions.ComputeGlobIgnoreCarriageReturn,
            generatorOptions.ComputeGlobIgnoreWhitespace,
            generalOptions,
            versions);

    public IReadOnlyDictionary<IInput, int> Provides() =>
        ((IInputProvider)this).DefaultPriorityMap(
            new IInput[]
            {
                Inputs.OriginGlob, Inputs.Transformations, Inputs.DestinationExcludePaths,
            });

    /// <summary>
    /// Resolves a destination path for glob generation heuristics. This allows the destination path
    /// to be different than the generator output folder, if needed. Port of the functional interface
    /// <c>DestinationPathProvider</c>.
    /// </summary>
    /// <exception cref="System.Threading.ThreadInterruptedException"/>
    /// <exception cref="CannotProvideException"/>
    public delegate string DestinationPathProvider(IInputProviderResolver db);
}
